enhancements/builds/multi-arch-builds.md

```diff
@@ -0,0 +1,456 @@
+* Extend BuildStatus to create a new type (BuildStatusPlatform?) which includes a platform specifier string, replacing its Output field with a BuildStatusPlatformOutput.
+* Add a []BuildStatusPlatform field to github.com/openshift/api/build/v1.BuildStatus. The platform values in the structs will be used to index them when patching the structure.
+* Ensure that github.com/openshift/client-go/build/clientset/versioned/typed/build/v1.builds.UpdateDetails() does not lose data if multiple client processes are calling it at roughly the same time.
+* Teach openshift-builder to observe the target platforms list in the encoded Build structure it's given, to relay it down to the buildah library, and to update the appropriate BuildStatusPlatform and BuildStatusPlatformOutput values for the platform-specific it is performing. If the list is empty, it will instead build for the executing node's platform and update the overall build status.
+* Implement the manifest bundling command for openshift-builder. Once all build pods have finished running, the BuildStatusPlatform slice's members' BuildStatusPlatformOutput will include references to the images which need to be in the produced manifest list, along with additional information which should be included to describe each entry in the manifest list.
+* Ensure that the `openshift-controller-manager-images` configMap in the `openshift-controller-manager-operator` namespace references builder and deployer images using manifest lists and not architecture-specific images.
+* Add an optional target platforms []string to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Add an optional "archNodeSelector" *bool to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Add an optional "manifestList" *bool to github.com/openshift/api/openshiftcontrolplane/v1.BuildOverridesConfig.
+* Figure out where the platforms list for builds should come from if the configured BuildOverridesConfig doesn't have one.
+* Teach openshift-controller-manager to launch, monitor, and when necessary cancel the N+1 pods, setting the target platform and output location differently in the encoded Build values it passes to the various platform-specific build pods, unless "manifestList" is explicitly disabled.
+  * Build pods will gain a label which `oc logs` can use to select which of the build's pods it will read logs from.
+  * If a node with the corresponding OS and architecture selector labels (e.g. "linux", "amd64") is not able to build for any requested variant (e.g. "", "v2", etc.), more refinement will be required.
+  * Unless BuildOverridesConfig.archNodeSelector is explicitly disabled, each pod will have a "kubernetes.io/arch" value added to its "nodeSelector".
```

Comment:
i'm wondering if it is also worth supporting addition of tolerations in case nodes are tainted. eg: an admin taints a node of a different architecture after adding it to the cluster to not disrupt existing workloads

Reply:
I'm not sure how that would look in practice.  Do you have an example in mind, or is this an it-could-be-future-proofing thing?

Reply:
Hi @nalind,

We don't need this to be automatic, but those users tainting second+ architecture nodes could need to add a list of tolerations in their BuildConfigs.

Tolerations are safe to be added even for builds that land on non-tainted nodes. 

Therefore, if you assume having non-tainted amd64 nodes and tainted arm64+z nodes, a build config like the following one could be scheduled correctly on all the architectures.

```yaml
kind: BuildConfig....
...
spec:
  nodeSelector:
    ... users' nodeSelectors, if any ...
  tolerations:
   - key: my-taint-on-arm64-nodes
     operator: Exists
     effect: NoSchedule
   - key: my-taint-on-z-nodes
     operator: Exists
     effect: NoSchedule
```

Reply:
Okay, then we'll need to add `Tolerations` to `github.com/openshift/api/build/v1.CommonSpec` and have the build controller pass the value through to the builder pods, as it does with `NodeSelector` now.